### PR TITLE
plugin: add audit plugin extension point

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

to support audit plugin

### What is changed and how it works?

add extension point and pass enough information into plugin.

### Check List 

Tests 

 - WIP

Code changes

 - Has exported function/method change
 - Has interface methods change

Side effects

 - N/A

Related changes

 - N/A

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/9136)

### CODE REVIEW

**Comment by lysu** [comment]
this pr also need https://github.com/pingcap/parser/pull/32 to mask insensitive data

File: executor/simple.go [hunk 1]
```diff
@@ -377,19 +377,16 @@ func (e *SimpleExec) executeFlush(s *ast.FlushStmt) error {
...
-		dom := domain.GetDomain(e.ctx)
-		if plugin.Get(plugin.Audit, "ipwhitelist") != nil {
-			if cli := dom.GetEtcdClient(); cli != nil {
-				const whitelistKey = "/tidb/plugins/whitelist"
-				row := cli.KV
-				_, err := row.Put(context.Background(), whitelistKey, "")
-				if err != nil {
-					log.Warn("notify update whitelist failed:", err)
-				}
+	case ast.FlushTiDBPlugin:
```
Comment (tiancaiamao) [inline executor/simple.go]: `FlushTiDBPlugin` is irrelevant to this PR, you can do it in a separate one.

File: plugin/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+WIP
```
Comment (tiancaiamao) [inline plugin/README.md]: ## The plugin framework

https://github.com/pingcap/tidb/blob/master/docs/design/2018-12-10-plugin-framework.md

File: plugin/plugin.go [hunk 3]
```diff
@@ -145,7 +150,7 @@ func (p *Plugin) validate(ctx context.Context, tiPlugins *plugins, mode validate
 
 // Init initializes the plugin and load plugin by config param.
 // This method isn't thread-safe and must be called before any other plugin operation.
-func Init(ctx context.Context, cfg Config) (err error) {
+func Init(ctx context.Context, dom *domain.Domain, cfg Config) (err error) {
```
Comment (tiancaiamao) [inline plugin/plugin.go]: So, plugin Init must be after domain initialize finish ?

File: executor/simple.go (hunk 1 above)
Comment (lysu) [reply executor/simple.go]: ok

File: server/conn.go [hunk 4]
```diff
@@ -712,6 +733,21 @@ func (cc *clientConn) dispatch(data []byte) error {
 		span.Finish()
 	}()
 
+	err := plugin.ForeachPlugin(plugin.Audit, func(p *plugin.Plugin) error {
+		auditPlugin := plugin.DeclareAuditManifest(p.Manifest)
+		if auditPlugin.OnGeneralEvent != nil {
+			sql := cc.lastCmd
+			if cmd == mysql.ComQuery {
+				sql = parser.Normalize(sql)
```
Comment (tiancaiamao) [inline server/conn.go]: If there are several Audit plugin, `parser.Normalize` is called multiple times.

File: plugin/audit.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	ClientPort        string
+	ServerID          int
+	ServerPort        int
+	Duration          int64
+	User              string
+	ProxyUser         string
+	PrivUser          string
+	ClientOsLoginUser string
+	ServerOsLoginUser string
+	OsVersion         string
```
Comment (tiancaiamao) [inline plugin/audit.go]: s/OsVersion/OSVersion

File: server/server.go [hunk 6]
```diff
@@ -55,6 +55,7 @@ import (
 	"github.com/pingcap/tidb/sessionctx/variable"
 	"github.com/pingcap/tidb/util"
 	log "github.com/sirupsen/logrus"
+	"os"
```
[the remark this answers is not in the document]
Comment (lysu) [reply server/server.go]: - - this will be lazy... WIP just a temp commit sorry- -

**Comment by tiancaiamao** [comment]
OK, I'll review it when it's ready.
Kind remind: don't make a PR too large

**Comment by lysu** [comment]
@tiancaiamao  ok~

**Comment by lysu** [comment]
@tiancaiamao @jackysp  PTAL when free thx~

File: server/tidb_test.go [hunk 7]
```diff
@@ -72,7 +72,7 @@ func (ts *TidbTestSuite) SetUpSuite(c *C) {
 
 	// Run this test here because parallel would affect the result of it.
 	runTestStmtCount(c)
-	defaultLoadDataBatchCnt = 3
+	defaultLoadDataBatchCnt = 1
```
Comment (tiancaiamao) [inline server/tidb_test.go]: Irrelevant change?

**Comment by tiancaiamao** [comment]
Rest LGTM
PTAL @jackysp

**Comment by tiancaiamao** [comment]
LGTM

**Comment by zz-jason** [comment]
@lysu CI failed.

**Pull request merged**